### DISCUSSION TITLE
Command line text editor

### DISCUSSION THREAD

**Discussion by jdescelliers** [PowerShell/PowerShell#16653] [Enhancements]
Hi,

While PS is quite useful for many tasks, I find that basic things like writing/editing a file is not as easy as it would be on unix/linux that are usually bundled with `vi` or `nano`.

I understand that it might not be the right place to this feature request, so there might be a better place for this feature request.

Maybe the old `edit.exe` would do the trick?

Is there any plans to provide such tool for Windows command line?

I believe `vi` is being able to handle very large files which could be a plus.

**Comment by 237dmitry** [comment]
In fact, there are modern console text editors. Try [micro](https://github.com/zyedidia/micro/releases) or the built-in [Far Manager](https://github.com/FarGroup/FarManager/releases) editor. Both of them have syntax highlighting, including powershell scripts.

**Comment by asmith3006** [comment]
Whilst I appreciate that there are installable ones available, a native one which you can rely on when you remote into any server (as long as the version of PS is up to date) would be much better. 
 
I cannot install apps I want on a remote server without a long approval process, so a built in one would be much more useful. 

I don't even care about syntax highlighting, just being able to change some text in remote config files would be a huge help.

**Comment by rkeithhill** [comment]
I like the suggestion, but it feels more like a suggestion for Windows instead of PowerShell.

**Reply by doctordns** [reply]
And for Linux. Although, in the Windows world, we have Notepad.exe (both of them) that come in the Windows folder. Not sure what possible value another text editor could possibly add.

**Reply by rkeithhill** [reply]
For Linux/mac I can edit files directly in a Windows Terminal ssh session using `vim`.  The OP is looking for something similar on Windows - like the old edit.com.  For Windows, since I always have Git installed, I add this to my profile:

```powershell
# Windows specific configuration
if ($IsWindows) {
    $env:PAGER = 'less -Ps"Page %db?B of %D:.\. Press h for help or Q to quit\."'
    Set-Alias nano 'C:\Program Files\Git\usr\bin\nano.exe'
    Set-Alias vim  'C:\Program Files\Git\usr\bin\vim.exe'
}
```

That way I can use `vim` (or nano) when I ssh into my Windows build machines.

**Reply by asmith3006** [reply]
> And for Linux. Although, in the Windows world, we have Notepad.exe (both of them) that come in the Windows folder. Not sure what possible value another text editor could possibly add.

You can't use notepad.exe from the command line. This is GUI, I'm suggesting TUI.

**Comment by asmith3006** [comment]
I disagree, it being universal on Win, Mac and Linux would be very nice in my opinion.

**Comment by kasini3000** [comment]
no ，no，no，
I like the suggestion, 
* It should be a powershell module in the PowerShell gallery. Only install module XXX is required to install it.
* It should be TUI interface and support the local language interface.
* It should be similar to edit.exe
* It does not require PowerShell syntax highlighting, but simply edits the text.
Because we can edit, debug PS1 scripts on remote Linux through vscode + Remote SSH

----

gui.CS framework can fully meet the requirements.
gui.CS for .net6,can run on win，linux，mac。
https://www.nuget.org/packages/Terminal.Gui/

Visit my 《power alias》
https://gitee.com/chuanjiao10/powershell/tree/master/win-linux%E9%80%9A%E7%94%A8/poweralias

Because I am just a beginner of GUI CS, I need someone to cooperate with me to develop this "PowerShell module", and I am responsible for the Chinese translation part.

**Comment by doctordns** [comment]
I am probably missing something - but if you want to edit PowerShell code - can't you use VS Code? See: [How to Setup VS Code on Linux](https://code.visualstudio.com/docs/setup/linux). If you are contributing to a shared source module, it would seem crazy to want to use any other tool - but I am fairly biased!

On Windows, in the absence of VS Code - you have Notepad (twice for good measure). And I would assume any even remotely component Linux IT Pro would either have gone over to the dark side (ie learned to be productive in VI) or found alternatives. There are [numerous alternatives to VI on Linux](https://www.tecmint.com/best-open-source-linux-text-editors). 

And finally, a text editor really seems outside the scope of the PowerShell team. If a new editor is needed, then Github, PSGallery, et al is that way ->.

**Reply by iSazonov** [reply]
They ask about TUI editor for pwsh.

**Reply by doctordns** [reply]
How many Markdown extensions are there for VS Code! :-)

**Reply by ThomasNieto** [reply]
If I'm understanding the ask correct, the use case would be to have a PS remoting friendly command-line text editor. On Linux vi can be used but it not installed by default on Windows. In the past I've wanted ways to easily edit files in a PS remoting session in Windows as easily as I can do in Linux with VIM.

I agree with the others on this thread that I don't think the PowerShell team has the resources or desire to create, maintain, and increase the size of the PS package for this functionality. However, I do think that gui.cs could be used to create a cross platform basic text editor and shipping that in a standalone module would meet this need.

**Reply by asmith3006** [reply]
Exactly as @iSazonov says. I'm not after an editor for PS files, I'm after an editor for ANY text files to make quick changes to. When I've remoted into a machine I want to edit a config file and there's no way to do this from the CLI on Windows without installing extra programs. It doesn't need syntax highlighting, line numbers, spell check or anything beyond open, edit, save. It'd be great to have find/replace, regex, but it's not essential.

**Reply by rkeithhill** [reply]
The past few years, Windows has added some *nix command line tools e.g.: scp, ssh/sshd, curl and tar. I wonder if they'd be open to adding vi/vim?

**Reply by SeeminglyScience** [reply]
> The past few years, Windows has added some *nix command line tools e.g.: scp, ssh/sshd, curl and tar. I wonder if they'd be open to adding vi/vim?

Would probably still need some support on the pwsh end to be able to run that in a winrm session. I put together a proof of concept using the same mechanism that `psedit` uses, but for vim [a few years ago here](https://github.com/PowerShell/PowerShell-RFC/issues/69#issuecomment-446217675).

**Reply by iSazonov** [reply]
> I wonder if they'd be open to adding vi/vim?

GNU license :-) blocks.
